File: algorithms/doublepointer/ispalindrome125/Solution.java
package blogandquestion.algorithms.doublepointer.ispalindrome125;

/**
 * @author Heming
 * created by IntelliJ IDEA
 * founded at 2020/10/15 22:22:06
 * description: 双指针，分别从左右两边开始验证。
 *
 * 给定一个字符串，验证它是否是回文串，只考虑字母和数字字符，可以忽略字母的大小写。
 *
 * 说明：本题中，我们将空字符串定义为有效的回文串。
 *
 * 示例 1:
 *
 * 输入: "A man, a plan, a canal: Panama"
 * 输出: true
 * 示例 2:
 *
 * 输入: "race a car"
 * 输出: false
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/valid-palindrome
 *
 */
public class Solution {
    public boolean isPalindrome(String s) {
        if (s == null) return false;
        else if (s.equals("")) return true;

        s = s.toLowerCase();
        int i = 0;
        int j = s.length() - 1;
        while (i < j) {
            char l =s.charAt(i);
            char r = s.charAt(j);
            // 判断字符是否是字母和数字(0 <= l - '0' <= 9, 数字)(49 <= l - '0' <= 74, 小写字母)
            // 如果不是字母和数字，直接跳过
            if (l - '0' < 0 || l - '0' > 9 && l - '0' < 49 || l - '0' > 74) {
                i++;
                continue;
            }
            if (r - '0' < 0 || r - '0' > 9 && r - '0' < 49 || r - '0' > 74) {
                j--;
                continue;
            }
            if (l != r) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        String s1 = "A man, a plan, a canal: Panama";
        boolean result = s.isPalindrome(s1);
        System.out.println(result);
    }
}

File: algorithms/doublepointer/ispalindrome125/solution.js
/*! *****************************************************************************
@author Heming
founded at 2020-11-29 10:45:05
created by IntelliJ IDEA
description: LeetCode 125：验证回文串
from：https://leetcode-cn.com/problems/valid-palindrome/
***************************************************************************** */

/**
 * @param {string} s
 * @return {boolean}
 */
const isPalindrome = function (s) {
  if (s.length === 0) return true;
  s = s.toLowerCase();
  let low = 0, high = s.length - 1;
  while (low < high) {
    if (!validChar(s[low])) {
      low++;
      continue;
    }
    if (!validChar(s[high])) {
      high--;
      continue;
    }
    if (s[low] !== s[high]) return false;
    low++;
    high--;
  }
  return true;

  /**
   * 如果字符c是小写字母或者数字，返回true，否则，返回false
   * @param c
   * @return {boolean}
   */
  function validChar(c) {
    let code = c.charCodeAt(0);
    return !(code < 48 || code > 57 && code < 97 || code > 122);
  }
};


/**
 * 没有上面执行的快😂
 * @param {string} s
 * @return {boolean}
 */
const isPalindromeReg = function (s) {
  if (s.length === 0) return true;
  s = s.toLowerCase();
  let low = 0, high = s.length - 1;
  let regexp = /[^0-9a-zA-Z]/;
  while (low < high) {
    // 使用正则表达式进行校验
    if (regexp.test(s[low])) {
      low++;
      continue;
    }
    if (regexp.test(s[high])) {
      high--;
      continue;
    }
    if (s[low] !== s[high]) {
      return false;
    }
    low++;
    high--;
  }
  return true;
}
